test/generate_expression_tests.py

```diff
@@ -0,0 +1,167 @@
+
+            # skip signatures without inputs that can be eigen types
+            if not sp.has_eigen_compatible_arg():
+                continue
+
+            # skip functions in forward mode with no forward mode overload, same for reverse mode
+            if overload == "Fwd" and not sp.is_fwd_compatible():
+                continue
+            if overload == "Rev" and not sp.is_rev_compatible():
+                continue
+
+            is_reverse_mode = overload == "Rev" and not sp.returns_int()
+
+            cg = CodeGenerator()
+
+            # Generate two sets of arguments, one will be expressions one will not
+            arg_list_expression_base = cg.build_arguments(sp, sp.number_arguments() * [overload], size = 1)
+            arg_list_expression = [cg.convert_to_expression(arg, size = 1) if arg.is_eigen_compatible() else arg for arg in arg_list_expression_base]
+            arg_list_no_expression = cg.build_arguments(sp, sp.number_arguments() * [overload], size = 1)
+
+            # Check results with expressions and without are the same
+            cpp_function_name = "stan::math::" + sp.function_name
+            result = cg.function_call_assign(cpp_function_name, *arg_list_expression)
+            result_no_expression = cg.function_call_assign(cpp_function_name, *arg_list_no_expression)
+            cg.expect_eq(result, result_no_expression)
+
+            # Check that expressions evaluated only once
+            for arg in arg_list_expression:
+                if arg.is_expression():
+                    cg.expect_leq_one(arg.counter)
+
+            # If reverse mode, check the adjoints
+            if is_reverse_mode:
+                summed_result = cg.recursive_sum(result)
+                summed_result_no_expression = cg.recursive_sum(result_no_expression)
+                sum_of_sums = cg.add(summed_result, summed_result_no_expression)
+                cg.grad(sum_of_sums)
+
+                for arg, arg_no_expression in zip(arg_list_no_expression, arg_list_expression):
+                    cg.expect_adj_eq(arg, arg_no_expression)
+
+                cg.recover_memory()
+
+            tests.append(
+                test_code_template.format(
+                    overload=overload,
+                    comment=signature.strip(),
+                    test_name=sp.function_name + repr(n),
+                    code = cg.cpp(),
+                )
+            )
+    
+    save_tests_in_files(j, tests)
+
+def processCLIArgs():
+    """
+    Define and process the command line interface to the benchmark.py script.
+    """
+    parser = FullErrorMsgParser(
+        description="Generate and run_command benchmarks.",
+        formatter_class=ArgumentDefaultsHelpFormatter,
+    )
+    parser.add_argument(
+        "--functions",
+        nargs="+",
+        type=str,
+        default=[],
+        help="Signatures and/or function names to benchmark. Ignores any finished checks in results file (if given).",
```

Comment:
I still don't think this file needs to be runnable on its own. If you strongly disagree, fix the help message and docs of `processCLIArgs`.

Reply:
Removed